File: main_auto.c
// main_auto.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/can.h>
#include <linux/can/raw.h>
#include <net/if.h>
#include <errno.h>
#include <pthread.h>
#include <time.h>

int sock;
struct timespec start_time;

void log_error(const char *msg) {
    FILE *log = fopen("main_log.txt", "a");
    if (log) {
        time_t now = time(NULL);
        fprintf(log, "[%s] ERROR: %s\n", strtok(ctime(&now), "\n"), msg);
        fclose(log);
    }
}

void log_info(const char *msg) {
    FILE *log = fopen("main_log.txt", "a");
    if (log) {
        time_t now = time(NULL);
        fprintf(log, "[%s] INFO: %s\n", strtok(ctime(&now), "\n"), msg);
        fclose(log);
    }
}

void setup_can_interface() {
    system("ip link set can0 down");
    system("ip link set can0 type can bitrate 50000");
    system("ip link set can0 up");
}

void *receive_data(void *arg) {
    struct can_frame frame;
    char buf[64] = {0};

    while (1) {
        if (read(sock, &frame, sizeof(frame)) > 0) {
            memcpy(buf, frame.data, frame.can_dlc);
            printf("\n[Received] From Module -> ID:%d -> %s\n", frame.can_id, buf);
            char log_msg[128];
            snprintf(log_msg, sizeof(log_msg), "Received from Module ID:%d -> %s", frame.can_id, buf);
            log_info(log_msg);
        }
    }
    return NULL;
}

int main() {
    setup_can_interface();
    clock_gettime(CLOCK_MONOTONIC, &start_time);

    struct sockaddr_can addr;
    struct ifreq ifr;
    struct can_frame frame;
    pthread_t recv_thread;

    if ((sock = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
        perror("Socket");
        log_error("Socket creation failed.");
        return 1;
    }

    strcpy(ifr.ifr_name, "can0");
    if (ioctl(sock, SIOCGIFINDEX, &ifr) < 0) {
        perror("ioctl");
        log_error("ioctl failed to get interface index.");
        return 1;
    }

    addr.can_family = AF_CAN;
    addr.can_ifindex = ifr.ifr_ifindex;

    if (bind(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("Bind");
        log_error("Socket bind failed.");
        return 1;
    }

    pthread_create(&recv_thread, NULL, receive_data, NULL);

    int module_ids[] = {11, 13, 59};
    int current = 0;

    while (1) {
        struct timespec now;
        clock_gettime(CLOCK_MONOTONIC, &now);
        int elapsed_ms = (now.tv_sec - start_time.tv_sec) * 1000 +
                         (now.tv_nsec - start_time.tv_nsec) / 1000000;

        if (elapsed_ms >= 10) {
            start_time = now; // Reset base time every 200ms

            int target = module_ids[current];
            char msg[8];
            snprintf(msg, sizeof(msg), "M%d", target);

            frame.can_id = target;
            frame.can_dlc = strlen(msg);
            memcpy(frame.data, msg, frame.can_dlc);

            if (write(sock, &frame, sizeof(frame)) != sizeof(frame)) {
                perror("Write");
                log_error("Failed to send CAN frame to module.");
            } else {
                printf("[Sent] To Module %d -> %s\n", target, msg);
                char log_msg[64];
                snprintf(log_msg, sizeof(log_msg), "Sent to Module %d -> %s", target, msg);
                log_info(log_msg);
            }

            current = (current + 1) % 3;
        }

        usleep(10000);  // 10ms check
    }

    close(sock);
    return 0;
}

File: main_auto.py
import os
import can
import time
import threading
from datetime import datetime

log_file = "main_log.txt"

def log_info(msg):
    with open(log_file, "a") as f:
        f.write(f"[{datetime.now().strftime('%a %b %d %H:%M:%S %Y')}] INFO: {msg}\n")

def log_error(msg):
    with open(log_file, "a") as f:
        f.write(f"[{datetime.now().strftime('%a %b %d %H:%M:%S %Y')}] ERROR: {msg}\n")

def setup_can_interface():
    os.system("ip link set can0 down")
    os.system("ip link set can0 type can bitrate 50000")
    os.system("ip link set can0 up")

def receive_data(bus):
    while True:
        try:
            msg = bus.recv()
            if msg is not None:
                data = msg.data.decode('utf-8', errors='ignore')
                print(f"\n[Received] From Module -> ID:{msg.arbitration_id} -> {data}")
                log_info(f"Received from Module ID:{msg.arbitration_id} -> {data}")
        except Exception as e:
            log_error(f"Receive Error: {str(e)}")

def main():
    setup_can_interface()
    time.sleep(0.5)  # Short delay after bringing up interface

    try:
        bus = can.interface.Bus(channel='can0', bustype='socketcan')
    except OSError as e:
        print("Socket creation failed:", e)
        log_error("Socket creation failed.")
        return

    threading.Thread(target=receive_data, args=(bus,), daemon=True).start()

    module_ids = [11, 13, 59]
    current = 0
    start_time = time.monotonic()

    while True:
        now = time.monotonic()
        elapsed_ms = (now - start_time) * 1000

        if elapsed_ms >= 100:
            start_time = now
            target = module_ids[current]
            msg_str = f"M{target}"
            msg_data = msg_str.encode('utf-8')

            try:
                msg = can.Message(arbitration_id=target, data=msg_data, is_extended_id=False)
                bus.send(msg)
                print(f"[Sent] To Module {target} -> {msg_str}")
                log_info(f"Sent to Module {target} -> {msg_str}")
            except can.CanError as e:
                print("Write failed:", e)
                log_error("Failed to send CAN frame to module.")

            current = (current + 1) % len(module_ids)

        time.sleep(0.01)  # 10ms sleep

if __name__ == "__main__":
    main()

